Mathlib/LinearAlgebra/FreeModule/PID.lean

```diff
@@ -440,6 +441,65 @@ structure Basis.SmithNormalForm (N : Submodule R M) (ι : Type*) (n : ℕ) where
   snf : ∀ i, (bN i : M) = a i • bM (f i)
 #align basis.smith_normal_form Basis.SmithNormalForm
 
+namespace Basis.SmithNormalForm
+
+variable {n : ℕ} {N : Submodule R M} (snf : Basis.SmithNormalForm N ι n) (m : N)
+
+lemma repr_eq_zero_of_nmem_range {i : ι} (hi : i ∉ Set.range snf.f) :
+    snf.bM.repr m i = 0 := by
+  obtain ⟨m, hm⟩ := m
+  obtain ⟨c, rfl⟩ := snf.bN.mem_submodule_iff.mp hm
+  replace hi : ∀ j, snf.f j ≠ i := by simpa using hi
+  simp [Finsupp.single_apply, hi, snf.snf]
+
+@[simp] lemma repr_comp_embedding_eq_smul :
+    snf.bM.repr m ∘ snf.f = (snf.bN.repr m : Fin n → R) • snf.a := by
+  ext j
+  obtain ⟨m, hm⟩ := m
+  obtain ⟨c, rfl⟩ := snf.bN.mem_submodule_iff.mp hm
+  replace hm : (⟨Finsupp.sum c fun i t ↦ t • (↑(snf.bN i) : M), hm⟩ : N) =
+      Finsupp.sum c fun i t ↦ t • ⟨snf.bN i, (snf.bN i).2⟩ := by ext; change _ = N.subtype _; simp
+  simp_rw [hm, LinearEquiv.map_finsupp_sum, map_smul, Subtype.coe_eta, repr_self,
+    Finsupp.smul_single, smul_eq_mul, mul_one, Finsupp.sum_single, Pi.mul_apply]
+  classical
+  simp_rw [snf.snf, map_smul, repr_self, Finsupp.smul_single, smul_eq_mul, mul_one,
+    Function.comp_apply, Finsupp.sum_apply, Finsupp.single_apply, EmbeddingLike.apply_eq_iff_eq,
+    Finsupp.sum_ite_eq', Finsupp.mem_support_iff, ite_not, ite_eq_right_iff]
+  exact fun a ↦ (mul_eq_zero_of_left a _).symm
+
+@[simp] lemma repr_apply_embedding_eq_repr_smul {i : Fin n} :
+    snf.bM.repr m (snf.f i) = snf.bN.repr (snf.a i • m) i := by
+  change (snf.bM.repr m ∘ snf.f) i = _
+  rw [repr_comp_embedding_eq_smul, Pi.smul_apply', smul_eq_mul, mul_comm _ (snf.a i), map_smul,
+    Finsupp.coe_smul, Pi.smul_apply, smul_eq_mul]
```

Comment:
I would suggest to put this result before `repr_comp_embedding_eq_smul`, so that the proof of `repr_comp_embedding_eq_smul` can look something like `by ext; simp [repr_apply_embedding_eq_repr_smul, mul_comm]` (hopefully!)